src/request/plan.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::{
+    backoff::Backoff,
+    pd::PdClient,
+    request::{KvRequest, Shardable},
+    stats::tikv_stats,
+    transaction::{resolve_locks, HasLocks},
+    Error, Result,
+};
+use async_trait::async_trait;
+use futures::{prelude::*, stream::StreamExt};
+use std::{marker::PhantomData, sync::Arc};
+use tikv_client_store::{HasError, HasRegionError, KvClient};
+
+/// A plan for how to execute a request. A user builds up a plan with various
+/// options, then exectutes it.
+#[async_trait]
+pub trait Plan: Sized + Clone + Sync + Send + 'static {
+    /// The ultimate result of executing the plan (should be a high-level type, not a GRPC response).
+    type Result: Send;
+
+    /// Execute the plan.
+    async fn execute(&self) -> Result<Self::Result>;
+}
+
+/// The simplest plan which just dispatches a request to a specific kv server.
+#[derive(Clone)]
+pub struct Dispatch<Req: KvRequest> {
+    pub request: Req,
+    pub kv_client: Option<Arc<dyn KvClient + Send + Sync>>,
+}
+
+#[async_trait]
+impl<Req: KvRequest> Plan for Dispatch<Req> {
+    type Result = Req::Response;
+
+    async fn execute(&self) -> Result<Self::Result> {
+        let stats = tikv_stats(self.request.label());
+        let result = self
+            .kv_client
+            .as_ref()
+            .expect("Unreachable: kv_client has not been initialised in Dispatch")
+            .dispatch(&self.request)
+            .await;
+        let result = stats.done(result);
+        result.map(|r| *r.downcast().expect("Downcast failed"))
+    }
+}
+
+pub struct MultiRegionPlan<P: Plan, PdC: PdClient> {
+    pub(super) inner: P,
+    pub pd_client: Arc<PdC>,
+}
+
+impl<P: Plan, PdC: PdClient> Clone for MultiRegionPlan<P, PdC> {
+    fn clone(&self) -> Self {
+        MultiRegionPlan {
+            inner: self.inner.clone(),
+            pd_client: self.pd_client.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl<P: Plan<Result: HasError> + Shardable, PdC: PdClient> Plan for MultiRegionPlan<P, PdC> {
+    type Result = Vec<Result<P::Result>>;
+
+    async fn execute(&self) -> Result<Self::Result> {
+        Ok(self
+            .inner
+            .shards(&self.pd_client)
+            .and_then(move |(shard, store)| async move {
+                let mut clone = self.inner.clone();
+                clone.apply_shard(shard, &store)?;
+                let mut response = clone.execute().await?;
+                match response.error() {
+                    Some(e) => Err(e),
+                    None => Ok(response),
+                }
+            })
+            .collect()
+            .await)
+    }
+}
+
+/// A technique for merging responses into a single result (with type `Out`).
+pub trait Merge<In>: Sized + Clone + Send + Sync + 'static {
+    type Out: Send;
+
+    fn merge(&self, input: Vec<Result<In>>) -> Result<Self::Out>;
+}
+
+#[derive(Clone)]
+pub struct MergeResponse<P: Plan, In, M: Merge<In>> {
```

Comment:
Some plans are named after "xxxPlan", some are "xxxResponse". Could we use a similar pattern for the names?